### ISSUE TITLE
mPAPI_register not running

### ISSUE THREAD

**Issue by mokra** [quepas/mPAPI#5]
First of all, huge THANKS for your efforts! I really need this for my research!
I'm using ubuntu 18, last octave and last papi version.
In octave, when running a test code 
`mPAPI_register({'FP_ARITH:SCALAR_SINGLE', 'L1D:REPLACEMENT', 'PAPI_L2_ICA'})` 
I'm getting this error:
```
error: mPAPI_register: mPAPI_register takes cell array with 1 to 0 counter descriptors (up to number of available hardware performance monitoring counters)
error: called from
     test at line 3 column 4
```
Please help me to get this running.
Thanks!

**Comment by quepas** [comment]
@mokra Hi !

From the error message it seems that there are no available performance counter registers in your CPU at the moment. If you look at the code `mPAPI_register.cpp`, the `0` in your error message comes from the variable [papi_num_counters](https://github.com/quepas/mPAPI/blob/master/mPAPI_register.cpp#L44) which is initialized as `int papi_num_counters = PAPI_get_opt(PAPI_MAX_HWCTRS, NULL);`. Basically, it takes the maximal number of available hardware counters.

There are several reasons for this, but first, let's validate that you have no access to performance counters. From your console type `./papi_component_avail`. The result is a summary of the PAPI library with `Number Hardware Counters` and if the value is `0` then this is the problem.

Possible solution:
1) Change the access privilege to the performance counters (more on [StackOverflow](https://stackoverflow.com/questions/32308175/papi-avail-no-events-available)):
```bash
sudo sh -c "echo 0 > /proc/sys/kernel/perf_event_paranoid"
```
Tell me if this helps.

**Comment by mokra** [comment]
Wow - That was fast! Thanks
Yes, you've nailed the problem - `Number Hardware Counters: 0` . I've run the code you suggested and voila - the number of counters is now `6` !
Now, if I run the code, Octave crashes - in the console it says  `octave exited with signal 6`.
It happens directly after I run the first code line calling `mPAPI_register`
What might it be?
I've run a strace - see the contnet of output:
[output2.txt](https://github.com/quepas/mPAPI/files/3353571/output2.txt)

**Comment by quepas** [comment]
My pleasure :)

Now it is getting harder! I couldn't reproduce the problem, but I was trying to do that on `Ubuntu 16.04`, `Octave 4.2.2` and `5.1`, and `PAPI 5.7.1.0`. I will try to reproduce it on `Ubuntu 18.04`, probably tomorrow. Unfortunately, I can't deduce anything from the stack trace.

For now, I can recommend you to go over the list of available native performance events with the command: `papi_native_avail > papi_native.txt` and to check if events `'FP_ARITH:SCALAR_SINGLE', 'L1D:REPLACEMENT'` are available. The third event is a preset event (they all start with the prefix `PAPI_`). You can check their availability with the command `papi_avail -a`.

In case if any of the events `'FP_ARITH:SCALAR_SINGLE', 'L1D:REPLACEMENT', 'PAPI_L2_ICA'` is missing, try to use the available events instead.

Good luck!

**Comment by mokra** [comment]
Thanks!
I can use other functions like mPAPI_enumNativeEvents, mPAPI_register({'PAPI_TOT_INS'}, 1234)
So it doesn't crash!
Can you help me perform a little test?
How do I count FLOPs for an ".m" File? So I want to run a MATLAB or Octave code and see how many FLOPs does it cost.
It it works, then everything is great!!!
Thanks in advance

**Comment by quepas** [comment]
Counting FLOPS depends on the MATLAB/Octave code, because loops perform scalar floating-point operations, whereas vectorized codes perform vector floating-point operations. The required steps include:

1) Find the __available__ performance event(s) counting floating-point operations. PAPI has [predefined events](https://icl.cs.utk.edu/projects/papi/wiki/PAPIC:PAPI_presets.3) for that purpose like: `PAPI_FP_INS`, `PAPI_FP_OPS` (not available on my computer). `PAPI_SP_OPS` counts single-precision operations and it scales for vector operations. Similarly, `PAPI_DP_OPS` counts double-precision operations. For counting only vector operations you can use `PAPI_VEC_SP` or `PAPI_VEC_DP`. Moreover, you could use native events, e.g. `FP_ARITH:128B_PACKED_DOUBLE` etc. if available.
2) Register the selected events and perform the test:
```matlab
ev = mPAPI_register({'PAPI_DP_OPS'});
mPAPI_tic(ev);
% Here goes your program
flops = mPAPI_toc(ev);
```

I can't find why the `mPAPI_register` fails for you. You could theoretically, use `mPAPI_register` and count FLOPS for a given thread. In the directory `common-matlab` function `findCompThreads()` will empirically find computational threads for you. However, this won't work well for multi-threaded execution.

Anyway, if I be able to reproduce the bug I will try to fix it.

**Comment by mokra** [comment]
My MATLAB Code contains loops, matrices, vector and matrix multiplications, bessel functions und eigenvalue methods with complex numbers.
In my case I do have access to `PAPI_FP_INS, PAPI_FP_OPS`,.
I am using `PAPI_DP_OPS` - It calculates the amount of FLOPs it seems. But every time it is a different amount. In fact, I've run the code for 20 times and ploted the FLOPs amount (for both events - `PAPI_FP_OPS` and `PAPI_DP_OPS`) - the deviation is not really small - see image attached
![Screenshot from 2019-07-03 13-21-31](https://user-images.githubusercontent.com/1849512/60587804-a5de2d80-9d95-11e9-935d-5edf4ec666fe.png)
.
Is it normal?
could you test it on your machine how many flops you get?
```
allflops=zeros(100,1); allflops2=zeros(100,1);  k=1; 
figure(1); clf;
evn='PAPI_DP_OPS';
% evn='PAPI_FP_OPS';
for k=1:20
    ev2 = mPAPI_register({evn});
    mPAPI_tic(ev2);
    
    s=200;
    km=zeros(s,s);
    for i=1:s
      km(i,i) = rand(1)*sqrt(-1);
    end
    [e,v]=eigs(km);
    flops = mPAPI_toc(ev2)
    allflops(k) = flops;
    allflops2(k) = flops;

    figure(1);
    plot(1:k-1, allflops(1:k-1));
    pause(.1);
end

```

**Comment by quepas** [comment]
I get a pretty "similar" results :) Before jumping to the answer you need to remember about a few issues.

1) _Performance counters are not always precise, nor deterministic_. See papers like: [Accuracy of performance counter measurements](http://sape.inf.usi.ch/sites/default/files/publication/ispass09.pdf), [Can hardware performance counters produce expected, deterministic results?](http://ash2.icl.utk.edu/files/publications/2010/icl-utk-451-2010.pdf). Recently, Intel is putting more and more effort to get them right. I think partially, because they have created performance models for their micro architectures like [_Top-down Microarchitecture Analysis Method_](https://sci-hub.tw/10.1109/ISPASS.2014.6844459) which is built using performance events. What I have seen so far is that since Skylake and beyond, the performance counters seem to be quite good. For much older micro architectures I have seen overcounting of flops by even `x10`.

2) _You need to pay attention to what events you are really measuring_. This statement is especially true for preset events `PAPI_` which supposed to be universal events. In reality, preset events are mapped to one or more native events available on your specific CPU model (you can lookup the mappings with the command `./papi_decode -a`). This means that `PAPI_DP_OPS` on one CPU might measure slightly different thing in comparison to the other CPU.
Moreover, modern CPU are very complex, they are superscalar (can execute many instructions at once), and speculative (can execute instruction in an execution path that might be taken in a short future). The speculative CPU means that the CPU issues and executes more instruction that are needed. Now, if you count issued/executed instructions instead of retired instructions (good executed instructions), then you will naturally overcount.

3) _You need to pay attention to what programs you are really measuring_. If you comment out this line `[e,v]=eigs(km);` then the results should be pretty steady. The reason is that underlying `eigs` algorithm is iterative which in numerical world might generate different rounding, different convergent path etc.

The answer is: yes, it is normal :)

**Comment by mokra** [comment]
Again thanks for that much effort!
Regarding 1 - I agree, that is pretty much dependent on the architecture. And at one point system X might need 10 flops to cal. one procedure and at another this might take 20 flops

Regarding 2 - This is confusing. Assuming I have 6 cores and nothing is running in the background - the results might differ by 50 or even 100% ?

Regarding 3 - with iterative function I agree. But what about some simple tests like:
```
a=100;
a+a+a;
```
this should stably return 3 flops with the event `PAPI_DP_OPS` - but it doesn't. sometimes I get 1000 flops. This part is very confusing- 

What I wish to do is to compare the computation cost in terms of FLOPs of 2 different codes (written by me an using eigs and other functions)
They might differ just by 40% - but if the accuracy is that "jumpy" it seems I wont be able to do that comparison with papi - do you agree?
thanks again

**Comment by quepas** [comment]
Not a problem! :)

2) Hard to tell by how much. Depending on what you are really measuring, the state of the machine/OS etc. etc.
3) This is because of the bias/systematic error. If you do the same computation, but for `a` being a vector of size `100000` then the precision increases (not necessary the accuracy, because of the bias).

Usually, the measurements are jumpy. What you could do is to measure both programs many times and compute confidence intervals (using bootstrapping, because we don't know the distribution of the measurements, and they are not normal imho -- they tend to be skewed). Finally, compare the confidence intervals. If they are disjoint, then you have a real difference there.

However, because modern CPUs exploit e.g. Instruction Level Parallelism (ILP), focusing solely on FLOPS might not be the best idea if you care about how much time your program will take. Consider the fact that back end of the CPU is capable of executing several floating-point operations at once, if there are no data dependencies in your programs (meaning, the computations could be run in parallel). I am talking about one thread execution here. Add to this parallel execution and the difference might be even bigger. My point is, you can have a program which executes more FLOPS than the other, but still executes faster on a given CPU. I didn't mention the memory bandwidth/latency bottlenecks etc. and many other factors.

In summary, many performance events are great to compare how given two programs execute on a given platform. But there is no one performance event that will tell you that :(

Check out the metrics used in [_Top-Down Micro architecture Model_](https://download.01.org/perfmon/TMA_Metrics.xlsx) (it was open-sourced by Intel). I have actually started to apply this model to performance analysis of MATLAB programs and it can give you interesting information like: your program is memory or computation bound and so on.

**Comment by mokra** [comment]
Well this really changes some things.
I always thought of floating point operations as the key importance but it seems they are not - in fact a group of factors are including the architecture. 
So summing it up, I could run a couple of computations, cut out the peaks (that are definitely some unwished results) and look at the Gauss-distribution (normal distribution). If the standard deviation is not too high, I might use this values. 
Thanks a lot for your help - I will go with this solution rather then using `tic, toc` - since this is very much unreliable.

**Comment by quepas** [comment]
No problem and good luck!
